echo-notifications/src/main/groovy/com/netflix/spinnaker/echo/controller/NotificationController.java

```diff
@@ -0,0 +1,111 @@
+  @Autowired(required = false)
+  Collection<NotificationService> notificationServices;
+
+  @Autowired private InteractiveNotificationCallbackHandler interactiveNotificationCallbackHandler;
+
+  @Autowired Optional<List<NotificationAgent>> notificationAgents;
+
+  /**
+   * Provides an endpoint for other Spinnaker services to send out notifications to users.
+   * Processing of the request is delegated to an implementation of {@link NotificationService}
+   * appropriate for the {@link Notification#notificationType}.
+   *
+   * @param notification The notification to be sent, in an echo-generic format.
+   * @return
+   */
+  @RequestMapping(method = RequestMethod.POST)
+  public EchoResponse create(@RequestBody Notification notification) {
+    val notificationService =
+        notificationServices.stream()
+            .filter(
+                it ->
+                    it.supportsType(notification.getNotificationType())
+                        &&
+                        // Only delegate interactive notifications to interactive notification
+                        // services, and vice-versa.
+                        // This allows us to support two implementations in parallel for the same
+                        // notification type.
+                        (it instanceof InteractiveNotificationService
+                            ? notification.isInteractive()
+                            : !notification.isInteractive()))
+            .findFirst();
+    return notificationService.map(ns -> ns.handle(notification)).orElse(null);
+  }
+
+  /**
+   * Provides a generic callback API for notification services to call, primarily in response to
+   * interactive user action (e.g. clicking a button in a message). This method makes as few
+   * assumptions as possible about the request, delegating the raw request headers, parameters and
+   * body to the corresponding [InteractiveNotificationService] to process, and similarly allows the
+   * notification service to return an arbitrary response body to the caller.
+   *
+   * <p>Note that this method must be exposed externally through gate so that external services
+   * (e.g. Slack) may call back to echo.
+   *
+   * @param source The unique ID of the calling notification service (e.g. "slack")
+   * @param headers The request headers
+   * @param rawBody The raw body of the request
+   * @param parameters The request parameters, parsed as a Map
+   * @return
+   */
+  @RequestMapping(
+      value = "/callbacks/{source}",
+      method = RequestMethod.POST,
+      consumes = MediaType.APPLICATION_FORM_URLENCODED_VALUE,
+      produces = MediaType.APPLICATION_JSON_VALUE)
+  public ResponseEntity<String> processCallback(
+      @PathVariable String source, RequestEntity<String> request) {
+    return interactiveNotificationCallbackHandler.processCallback(source, request);
+  }
+
+  @GetMapping("/metadata")
+  public List<NotificationAgent> getNotificationTypeMetadata() {
```

Comment:
This handler should be the only difference between the old Groovy controller and this file.

Reply:
Thanks for converting this.